### PR TITLE
Ports Minesweeper

### PR DESCRIPTION
**Pull request by Majkl-J**
## About The Pull Request

Ports yogstation's minesweeper app and arcade machine to Bubbers.

I had to bite through about 70 errors but from what testing I have done both the backend and frontend are working as intended.
Will add proper emagging of the arcade machine in a future PR after this is confirmed to not crash on some magic tgui error I may have missed

## Changelog

:cl:
add: Added minesweeper to PDAs and also machines
/:cl:

### CODE REVIEW

**Comment by nevimer** [comment]
Have it pass CI and I will testmerge it

**Comment by Majkl-J** [comment]
Turns out they killed get_header_data at the worst possible moment

**Comment by Majkl-J** [comment]
image
Ok I hold no responsibility for this runtime how did this even appear

**Comment by nevimer** [comment]
> image Ok I hold no responsibility for this runtime how did this even appear

It's a flakey issue, no worries.

**Comment by Majkl-J** [comment]
Well uh that's all that I may have caused fixed, no clue what the last runtime is, probably not my fault

File: modular_zubbers/code/game/machinery/computer/arcade.dm [hunk 1]
```diff
@@ -0,0 +1,157 @@
+#define MINESWEEPER_BEGINNER 1
```
Comment (nevimer) [inline modular_zubbers/code/game/machinery/computer/arcade.dm:1]: Move these defines into the file that uses them, and add #undef at the end of the file that uses them. minesweeper.dm seems to be the only the uses them.

File: modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+	game_status = result
+	if(result == MINESWEEPER_VICTORY)
+		play_snd('modular_zubbers/sound/arcade/minesweeper_win.ogg')
+		host.say("You cleared the board of all mines! Congratulations!")
+		if(emaggable && host.obj_flags & EMAGGED && value >= 1)
+			var/itemname
+			switch(rand(1,3))
+				if(1)
+					itemname = "a syndicate bomb beacon"
+					new /obj/item/sbeacondrop/bomb(host.loc)
```
Comment (nevimer) [inline modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm:263]: we might need to talk about the rewards

File: modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+				if(2)
+					itemname = "a rocket launcher"
```
Comment (nevimer) [inline modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm:265]: oh no, oh nooooo

File: modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+				if(3)
+					itemname = "two bags of c4"
+					new /obj/item/storage/backpack/duffelbag/syndie/c4(host.loc)
+					new /obj/item/storage/backpack/duffelbag/syndie/x4(host.loc)
```
Comment (nevimer) [inline modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm:269]: damn one emag for two bags of explosives that is crazy value

File: modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+		if(emaggable && host.obj_flags & EMAGGED && value >= 1)
+			var/itemname
+			switch(rand(1,3))
+				if(1)
+					itemname = "a syndicate bomb beacon"
+					new /obj/item/sbeacondrop/bomb(host.loc)
+				if(2)
+					itemname = "a rocket launcher"
+				if(3)
+					itemname = "two bags of c4"
```
Comment (nevimer) [inline modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm:267]: two c4 or one of both?

File: modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+		for(var/scany=-1, scany<2, scany++)
+			if(scanx+x < 1 || scany+y < 1)
+				continue
+			var/list/C = list(scanx+x-1,scany+y-1)
+			if(coord[1] == C[1] && coord[2] == C[2])
+				return TRUE
+	return FALSE
+
+/datum/minesweeper/proc/KABLOOEY()
+	explosion(get_turf(host),1,3,rand(1,5),rand(1,10))
```
Comment (nevimer) [inline modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm:412]: this is a rather devastating boom

File: modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+/datum/minesweeper/proc/find_in_mines(list/coord)
+	var/order = 0
+	for(var/list/L in mine_spots)
+		order++
+		if(coord[1] == L[1] && coord[2] == L[2])
+			return order
+	return FALSE
+
+/datum/minesweeper/proc/is_blank_tile_start(x,y)
+	for(var/scanx=-1, scanx<2, scanx++) // -1, 0, 1
```
Comment (nevimer) [inline modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm:381]: theres a microopt to be had here

File: modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+			if(scanx+x < 1 || scany+y < 1)
+				continue
+			if(scanx == 0 && scany == 0) // We know we aren't a mine
+				continue
+			if(board_data[scanx+x][scany+y] != "minesweeper_hidden.png" && board_data[scanx+x][scany+y] != "minesweeper_flag.png")
+				continue
+			if(find_in_mines(list(scanx+x-1,scany+y-1)))
+				mine_count++
+
+	if(mine_count == 0) // There are no mines around me! Select every square adjacent!
```
Comment (nevimer) [inline modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm:353]: falsey might be better

**Review by nevimer — CHANGES_REQUESTED** [review]
Put a video in the PR body of it running and working too

**Comment by Majkl-J** [comment]
I may have forgotten to cut those out woopsie

File: modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm (hunk 7 above)
Comment (Majkl-J) [reply modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm:381]: No clue what this is meant to be about here

**Comment by Majkl-J** [comment]
# The video


https://github.com/Bubberstation/Bubberstation/assets/49160555/54a66d6d-9775-4fec-8ec1-0c881509ce49

**Comment by nevimer** [comment]
update your branch to master

File: modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm (hunk 7 above)
Comment (Majkl-J) [reply modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm:366]: The fuck is a microopt

Comment (nevimer) [reply modular_zubbers/code/modules/modular_computers/file_system/programs/minesweeper.dm:366]:
```
for(var/i=-1 to limit)
    dothing()
```

runs faster than a traditional javascript style loop

**Pull request merged**